### ISSUE TITLE
wgpu::Queue::write_texture (destination and source slices have different lengths)

### ISSUE THREAD

**Issue by bjrnmrtns** [gfx-rs/wgpu-rs#572]
When using write_texture I get the following compile error: destination and source slices have different lengths

This happens when I don't write the full source slice.

The numbers the source size which is shown is also exactly the part I want to copy.

Here is the code I use to copy a part of the slice:
`
let data: Vec<f32> = vec!(1.0; 256);
queue.write_texture(wgpu::TextureCopyView {
                    texture: &self.texture,
                    mip_level: 0,
                    origin: wgpu::Origin3d {
                        x: 0,
                        y: 0,
                        z: 0
                    },
                },
                bytemuck::cast_slice(data.as_slice()),
wgpu::TextureDataLayout {
           offset: 4 as wgpu::BufferAddress,
           bytes_per_row: 16 * 4,
           rows_per_image: 16,
}, 
wgpu::Extent3d {
         width: 4 as u32,
         height: 4 as u32,
         depth: 1
 });
`

It seems correct code, but I'm not sure. Is the slice check maybe wrong? Currently it stops me from partly copying a buffer in this way. Is this correct?

**Comment by kvark** [comment]
You are copying 4 rows of 64 bytes each. So it expects exactly 256 bytes of data. What you are passing is 256 floats bitcasted to bytes, so that would be 1024 bytes.

**Comment by bjrnmrtns** [comment]
Yes what you are saying is correct, but my expectation from the interface is a bit different (maybe wrong).
What I expect is that I can pass in a buffer of 1024 bytes, and partly copy that.
This is how copy_buffer_to_texture works. I can give an offset into a Buffer and copy partly.
So if my expectation is wrong, why is there a copy_size: Extend3d? Because, from the layout and the destination you could derive everything you need to copy.
Am I missing something?

**Comment by kvark** [comment]
Yeah, we could totally use a part of the slice there. At the same time, this is rust, and taking a sub-slice of a slice is a trivial thing to do on your side as well. Having this more strict and explicit doesn't harm too much. At least you know exactly what part of the data you passed is used.
Also, I believe the `offset` is not used when calling `write_texture`. We need to have some validation on that.

**Comment by bjrnmrtns** [comment]
Taking a sub-slice is easy, but in a case where you want to copy a sub region of "2d data" it is not possible. Lets say I have a 4 by 4 2d array which ends up to be 16 elements, and I want to copy row 2 and 3 from column 2 until 3, that would mean I need to have a view on the data which ends up to be elements 5, 6, 8, 9.
Looking at the interface of write_texture, I would say, this looks possible.

**Comment by kvark** [comment]
In this case, your slice would see elements 5, 6, 7, 8, 9, 10 (a contiguous sub-slice).
It would first read 5, 6 of the row 2. Then it would want to advance to the next row by adding `bytes_per_row` to the base address, which would point it to element 9 (and skip 7 and 8), and then it would read elements 9 and 10 for the row 3.

**Comment by bjrnmrtns** [comment]
Ok if that works, that would be nice. I will check it out. Thanks.

**Comment by kvark** [comment]
So you are saying the offset should be respected? I'm totally open for that. Currently, it's awkward to ignore it.

**Comment by bjrnmrtns** [comment]
Yes I think so. I think you should be able to put in any arbitrary sized slice, and obviously when there is an access violation it is wrong, but otherwise do what de TextureCopyView TextureDataLayout and Extend3d describe.
Actually when using copy_buffer_to_texture (on the encoder), it works like that, but then you are copying from a wgpu::Buffer.

**Comment by kvark** [comment]
`copy_buffer_to_texture` is a bit different, because you don't provide a slice in there, just an offset in a buffer. So naturally there is a cut-off happening.

As for providing arbitrary-sized slices, we had a discussion in https://github.com/gpuweb/gpuweb/pull/903. If I remember correctly, the outcome was that you aren't supposed to pass down more data than needed by the copy.

**Comment by bjrnmrtns** [comment]
> As for providing arbitrary-sized slices, we had a discussion in gpuweb/gpuweb#903. If I remember correctly, the outcome was that you aren't supposed to pass down more data than needed by the copy.
I can't get why it is an issue to pass down more than needed by the copy.
What matters imho is
 -  not going outside bounds of the sources buffer
 - not going outside bounds of the target buffer
 - implementation should just follow what the arguments describe
Those are in my opinion the most flexible and the least surprising. Unless of course this is hard to implement or i'm still missing something.

**Comment by kvark** [comment]
Essentially, you are saying that my proposal in https://github.com/gpuweb/gpuweb/pull/903 makes more sense. I can't disagree :) But apparently, Google thinks differently.
